### ISSUE TITLE
Metalsmith fails with increased node count

### ISSUE THREAD

**Issue by meganhkelley** [department-of-veterans-affairs/va.gov-team#19232]
## Details

Notes from 1/28/21 standup: 
* Metalsmith layout plugin - in Jenkins (bc Jenkins is slow? maybe?) - is V1.8something, current version is V2.3, older versions load all files async. As it processes they drop off but if it tries to process too many things at once it fails and the error is “too many open files”
* We’re running with test data which is an increased node count
* How to resolve this? 
   * Upgrade the layout plugin? This would be a big effort bc it is a major version upgrade. And there’s no guarantee that that solves the problem. 
   * Could try using a graceful file system node module - less work but not a guarantee either. If the leak is coming from metalsmith plugin you can’t fix w/o upgrading. 
* Why is this not a problem using GraphQL? Theoretically bc GraphQL comes in in one file and CMS Export is thousands of files
* Because of this, cannot get Jenkins times for CMS Export with higher node count bc it breaks 

Additional details in Slack: 
https://dsva.slack.com/archives/CQH357ZTP/p1611860352060100

Goal of this ticket is to investigate why this is happening and maybe try a quick fix. If there is no quick fix, determine a solution approach and create separate tickets for implementing that solution.

**Comment by mchelen-gov** [comment]
CMS export completes in ~12m locally with 4096 nodes and fails on Jenkins https://docs.google.com/spreadsheets/d/1enwIGvEyGLVCLIHeRWXrp-ggDsnss8dyxSK4Bvxx4Mw/edit#gid=105374863

**Comment by meganhkelley** [comment]
Standup notes:

Tim: looked into ulimit increases that we discussed with Dror, Jenkins isn’t allowing it. Problem does occur in metalsmith, there’s a feature for it called concurrency to decrease the number of files processed. 
If that doesn’t work, will try to increase limit in Jenkins 
If that doesn’t work, Ops may be able to increase ulimit for us 
Potential risk with ulimit apporach — if we just increase the limit, we are going to hit the limit again 

Katy: followed up on using graceful-fs instead of fs for opening files. Dug into that and it did not work. 
Looking into: if we update the metalsmith layouts plugin, would it resolve the issue? This is throwing errors that we need to troubleshoot.

**Comment by k80bowman** [comment]
I was able to get the metalsmith-layouts plugin updated and working this morning, but it did not resolve the issue.

On Friday during the mob session we figured out that we can use the Activity Monitor to see what files are open during the node process. Using that, we were able to see that at points during the Apply layouts step in the build (step 20), over 3,000 files could be open at one time depending on how quickly the metalsmith-layouts plugin was handling them. I've added a couple of screenshots here to show what that looks like and what some of the files are that are being opened at this stage.

![Screen Shot 2021-02-01 at 1.30.03 PM.png](https://images.zenhubusercontent.com/5952ac1861b4ca3b0a727ca8/7a06b9b1-9394-4bcc-9e89-e7016d4e5a02)

![Screen Shot 2021-02-01 at 1.30.09 PM.png](https://images.zenhubusercontent.com/5952ac1861b4ca3b0a727ca8/ae48f883-973a-4b49-afa8-67b8983a8232)

**Comment by k80bowman** [comment]
@timwright12 was able to upgrade the ulimit on Jenkins through our Docker specifications ([PR](https://github.com/department-of-veterans-affairs/vets-website/pull/15858)) and this addressed the issue we were seeing with Metalsmith failing. This change has been merged into master already. @timwright12 has also validated that this fix remains valid through the 12,269 nodes we were able to test as part of the [CMS Export scalability validation](https://app.zenhub.com/workspaces/vsp-5cedc9cce6e3335dc5a49fc4/issues/department-of-veterans-affairs/va.gov-team/18592) work.

There is the possibility that the ulimit may need to be increased further in the future, but when this was discussed in a meeting about the CMS scalability findings (2/5/21), this did not seem to be a concern. If we need to increase the ulimit again in the future, we should be able to with little difficulty.

I think it's OK to call this issue resolved and close this card.
